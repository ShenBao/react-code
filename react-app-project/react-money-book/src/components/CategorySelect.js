import React from 'react'
import Ionicon from 'react-ionicons'
import PropTypes from 'prop-types'
import { Colors } from '../utility'

class CategorySelect extends React.Component {
  selectCategory = (event, category) => {
    this.props.onSelectCategory(category)
    event.preventDefault()
  }
  render() {
    const { categories, selectedCategory } = this.props
    const selectedCategoryId = selectedCategory && selectedCategory.id
    return (
      <div className="category-select-component">
        <div className="row">
        {
          categories.map((category, index) => {
            const iconColor = (category.id === selectedCategoryId) ? Colors.white : Colors.gray
            const backColor = (category.id === selectedCategoryId) ? Colors.blue : Colors.lightGray
            const activeClassName = (selectedCategoryId === category.id)
            ? 'category-item col-3 active' : 'category-item col-3'
            return (
              <div className={activeClassName} key={index} role="button" style={{ textAlign: 'center'}}
              onClick={(event) => {this.selectCategory(event, category)}}>
                <Ionicon
                  className="rounded-circle"
                  style={{ backgroundColor: backColor, padding: '5px' }} 
                  fontSize="50px"
                  color={iconColor}
                  icon={category.iconName}
                />
                <p>{category.name}</p>
              </div>
            )
          })
        }
        </div>
      </div>
    )
  }
}

CategorySelect.propTypes = {
  categories: PropTypes.array.isRequired,
  selectedCategory: PropTypes.object,
  onSelectCategory: PropTypes.func.isRequired,
}
export default CategorySelect