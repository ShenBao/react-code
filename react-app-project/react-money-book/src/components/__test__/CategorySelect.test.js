import React from 'react'
import { shallow } from 'enzyme'
import CategorySelect from '../CategorySelect'
import Ionicon from 'react-ionicons'

export const categories = [
   {
    "id": "1",
    "name": "旅行",
    "type": "outcome",
    "iconName": "ios-plane",    
  },
   {
    "id": "2",
    "name": "理财",
    "type": "income",
    "iconName": "logo-yen", 
  },
  {
    "id": "3",
    "name": "理财",
    "type": "income",
    "iconName": "logo-yen", 
  }
]

let props = {
  categories,
  onSelectCategory: jest.fn()
}

let props_with_category = {
  categories,
  onSelectCategory: jest.fn(),
  selectedCategory: categories[0],
}

describe('test CategorySelect component', () => {
  it('should render the component to match the snapshot', () => {
    const wrapper = shallow(<CategorySelect {...props_with_category} />)
    expect(wrapper).toMatchSnapshot()
  })
  it('renders with categories should render the correct items', () => {
    const wrapper = shallow(<CategorySelect {...props} />)
    expect(wrapper.find('.category-item').length).toEqual(categories.length)
    expect(wrapper.find('.category-item.active').length).toEqual(0)
    const firstIcon = wrapper.find('.category-item').first().find(Ionicon)
    expect(firstIcon.length).toEqual(1)
    expect(firstIcon.props().icon).toEqual(categories[0].iconName)
  })
  it('render selectedCategory with category item with highlight', () => {
    const wrapper = shallow(<CategorySelect {...props_with_category} />)
    expect(wrapper.find('.category-item').first().hasClass('active')).toEqual(true)
  })
  it('click the item should add active class and trigger the callback', () => {
    const wrapper = shallow(<CategorySelect {...props_with_category} />)
    wrapper.find('.category-item').at(1).simulate('click', { preventDefault: () => {} })
    expect(props_with_category.onSelectCategory).toHaveBeenCalledWith(categories[1])
  })
})

