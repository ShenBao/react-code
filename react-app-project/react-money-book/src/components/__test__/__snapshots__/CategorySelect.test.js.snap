// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test CategorySelect component should render the component to match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CategorySelect
    categories={
        Array [
            Object {
              "iconName": "ios-plane",
              "id": "1",
              "name": "旅行",
              "type": "outcome",
            },
            Object {
              "iconName": "logo-yen",
              "id": "2",
              "name": "理财",
              "type": "income",
            },
            Object {
              "iconName": "logo-yen",
              "id": "3",
              "name": "理财",
              "type": "income",
            },
          ]
    }
    onSelectCategory={[Function]}
    selectedCategory={
        Object {
            "iconName": "ios-plane",
            "id": "1",
            "name": "旅行",
            "type": "outcome",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="row"
>
        <div
                className="category-item col-3 active"
                onClick={[Function]}
                role="button"
                style={
                        Object {
                                "textAlign": "center",
                              }
                }
        >
                <Ionicon
                        beat={false}
                        className="rounded-circle"
                        color="#fff"
                        fontSize="50px"
                        icon="ios-plane"
                        rotate={false}
                        shake={false}
                        style={
                                Object {
                                        "backgroundColor": "#347eff",
                                        "padding": "5px",
                                      }
                        }
                />
                <p>
                        旅行
                </p>
        </div>
        <div
                className="category-item col-3"
                onClick={[Function]}
                role="button"
                style={
                        Object {
                                "textAlign": "center",
                              }
                }
        >
                <Ionicon
                        beat={false}
                        className="rounded-circle"
                        color="#555"
                        fontSize="50px"
                        icon="logo-yen"
                        rotate={false}
                        shake={false}
                        style={
                                Object {
                                        "backgroundColor": "#efefef",
                                        "padding": "5px",
                                      }
                        }
                />
                <p>
                        理财
                </p>
        </div>
        <div
                className="category-item col-3"
                onClick={[Function]}
                role="button"
                style={
                        Object {
                                "textAlign": "center",
                              }
                }
        >
                <Ionicon
                        beat={false}
                        className="rounded-circle"
                        color="#555"
                        fontSize="50px"
                        icon="logo-yen"
                        rotate={false}
                        shake={false}
                        style={
                                Object {
                                        "backgroundColor": "#efefef",
                                        "padding": "5px",
                                      }
                        }
                />
                <p>
                        理财
                </p>
        </div>
</div>,
      "className": "category-select-component",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="category-item col-3 active"
            onClick={[Function]}
            role="button"
            style={
                        Object {
                                    "textAlign": "center",
                                  }
            }
>
            <Ionicon
                        beat={false}
                        className="rounded-circle"
                        color="#fff"
                        fontSize="50px"
                        icon="ios-plane"
                        rotate={false}
                        shake={false}
                        style={
                                    Object {
                                                "backgroundColor": "#347eff",
                                                "padding": "5px",
                                              }
                        }
            />
            <p>
                        旅行
            </p>
</div>,
          <div
            className="category-item col-3"
            onClick={[Function]}
            role="button"
            style={
                        Object {
                                    "textAlign": "center",
                                  }
            }
>
            <Ionicon
                        beat={false}
                        className="rounded-circle"
                        color="#555"
                        fontSize="50px"
                        icon="logo-yen"
                        rotate={false}
                        shake={false}
                        style={
                                    Object {
                                                "backgroundColor": "#efefef",
                                                "padding": "5px",
                                              }
                        }
            />
            <p>
                        理财
            </p>
</div>,
          <div
            className="category-item col-3"
            onClick={[Function]}
            role="button"
            style={
                        Object {
                                    "textAlign": "center",
                                  }
            }
>
            <Ionicon
                        beat={false}
                        className="rounded-circle"
                        color="#555"
                        fontSize="50px"
                        icon="logo-yen"
                        rotate={false}
                        shake={false}
                        style={
                                    Object {
                                                "backgroundColor": "#efefef",
                                                "padding": "5px",
                                              }
                        }
            />
            <p>
                        理财
            </p>
</div>,
        ],
        "className": "row",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "0",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Ionicon
                beat={false}
                className="rounded-circle"
                color="#fff"
                fontSize="50px"
                icon="ios-plane"
                rotate={false}
                shake={false}
                style={
                                Object {
                                                "backgroundColor": "#347eff",
                                                "padding": "5px",
                                              }
                }
/>,
              <p>
                旅行
</p>,
            ],
            "className": "category-item col-3 active",
            "onClick": [Function],
            "role": "button",
            "style": Object {
              "textAlign": "center",
            },
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "beat": false,
                "className": "rounded-circle",
                "color": "#fff",
                "fontSize": "50px",
                "icon": "ios-plane",
                "rotate": false,
                "shake": false,
                "style": Object {
                  "backgroundColor": "#347eff",
                  "padding": "5px",
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "旅行",
              },
              "ref": null,
              "rendered": "旅行",
              "type": "p",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Ionicon
                beat={false}
                className="rounded-circle"
                color="#555"
                fontSize="50px"
                icon="logo-yen"
                rotate={false}
                shake={false}
                style={
                                Object {
                                                "backgroundColor": "#efefef",
                                                "padding": "5px",
                                              }
                }
/>,
              <p>
                理财
</p>,
            ],
            "className": "category-item col-3",
            "onClick": [Function],
            "role": "button",
            "style": Object {
              "textAlign": "center",
            },
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "beat": false,
                "className": "rounded-circle",
                "color": "#555",
                "fontSize": "50px",
                "icon": "logo-yen",
                "rotate": false,
                "shake": false,
                "style": Object {
                  "backgroundColor": "#efefef",
                  "padding": "5px",
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "理财",
              },
              "ref": null,
              "rendered": "理财",
              "type": "p",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Ionicon
                beat={false}
                className="rounded-circle"
                color="#555"
                fontSize="50px"
                icon="logo-yen"
                rotate={false}
                shake={false}
                style={
                                Object {
                                                "backgroundColor": "#efefef",
                                                "padding": "5px",
                                              }
                }
/>,
              <p>
                理财
</p>,
            ],
            "className": "category-item col-3",
            "onClick": [Function],
            "role": "button",
            "style": Object {
              "textAlign": "center",
            },
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "beat": false,
                "className": "rounded-circle",
                "color": "#555",
                "fontSize": "50px",
                "icon": "logo-yen",
                "rotate": false,
                "shake": false,
                "style": Object {
                  "backgroundColor": "#efefef",
                  "padding": "5px",
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "理财",
              },
              "ref": null,
              "rendered": "理财",
              "type": "p",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="row"
>
          <div
                    className="category-item col-3 active"
                    onClick={[Function]}
                    role="button"
                    style={
                              Object {
                                        "textAlign": "center",
                                      }
                    }
          >
                    <Ionicon
                              beat={false}
                              className="rounded-circle"
                              color="#fff"
                              fontSize="50px"
                              icon="ios-plane"
                              rotate={false}
                              shake={false}
                              style={
                                        Object {
                                                  "backgroundColor": "#347eff",
                                                  "padding": "5px",
                                                }
                              }
                    />
                    <p>
                              旅行
                    </p>
          </div>
          <div
                    className="category-item col-3"
                    onClick={[Function]}
                    role="button"
                    style={
                              Object {
                                        "textAlign": "center",
                                      }
                    }
          >
                    <Ionicon
                              beat={false}
                              className="rounded-circle"
                              color="#555"
                              fontSize="50px"
                              icon="logo-yen"
                              rotate={false}
                              shake={false}
                              style={
                                        Object {
                                                  "backgroundColor": "#efefef",
                                                  "padding": "5px",
                                                }
                              }
                    />
                    <p>
                              理财
                    </p>
          </div>
          <div
                    className="category-item col-3"
                    onClick={[Function]}
                    role="button"
                    style={
                              Object {
                                        "textAlign": "center",
                                      }
                    }
          >
                    <Ionicon
                              beat={false}
                              className="rounded-circle"
                              color="#555"
                              fontSize="50px"
                              icon="logo-yen"
                              rotate={false}
                              shake={false}
                              style={
                                        Object {
                                                  "backgroundColor": "#efefef",
                                                  "padding": "5px",
                                                }
                              }
                    />
                    <p>
                              理财
                    </p>
          </div>
</div>,
        "className": "category-select-component",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="category-item col-3 active"
              onClick={[Function]}
              role="button"
              style={
                            Object {
                                          "textAlign": "center",
                                        }
              }
>
              <Ionicon
                            beat={false}
                            className="rounded-circle"
                            color="#fff"
                            fontSize="50px"
                            icon="ios-plane"
                            rotate={false}
                            shake={false}
                            style={
                                          Object {
                                                        "backgroundColor": "#347eff",
                                                        "padding": "5px",
                                                      }
                            }
              />
              <p>
                            旅行
              </p>
</div>,
            <div
              className="category-item col-3"
              onClick={[Function]}
              role="button"
              style={
                            Object {
                                          "textAlign": "center",
                                        }
              }
>
              <Ionicon
                            beat={false}
                            className="rounded-circle"
                            color="#555"
                            fontSize="50px"
                            icon="logo-yen"
                            rotate={false}
                            shake={false}
                            style={
                                          Object {
                                                        "backgroundColor": "#efefef",
                                                        "padding": "5px",
                                                      }
                            }
              />
              <p>
                            理财
              </p>
</div>,
            <div
              className="category-item col-3"
              onClick={[Function]}
              role="button"
              style={
                            Object {
                                          "textAlign": "center",
                                        }
              }
>
              <Ionicon
                            beat={false}
                            className="rounded-circle"
                            color="#555"
                            fontSize="50px"
                            icon="logo-yen"
                            rotate={false}
                            shake={false}
                            style={
                                          Object {
                                                        "backgroundColor": "#efefef",
                                                        "padding": "5px",
                                                      }
                            }
              />
              <p>
                            理财
              </p>
</div>,
          ],
          "className": "row",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "0",
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <Ionicon
                  beat={false}
                  className="rounded-circle"
                  color="#fff"
                  fontSize="50px"
                  icon="ios-plane"
                  rotate={false}
                  shake={false}
                  style={
                                    Object {
                                                      "backgroundColor": "#347eff",
                                                      "padding": "5px",
                                                    }
                  }
/>,
                <p>
                  旅行
</p>,
              ],
              "className": "category-item col-3 active",
              "onClick": [Function],
              "role": "button",
              "style": Object {
                "textAlign": "center",
              },
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "beat": false,
                  "className": "rounded-circle",
                  "color": "#fff",
                  "fontSize": "50px",
                  "icon": "ios-plane",
                  "rotate": false,
                  "shake": false,
                  "style": Object {
                    "backgroundColor": "#347eff",
                    "padding": "5px",
                  },
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "旅行",
                },
                "ref": null,
                "rendered": "旅行",
                "type": "p",
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": "1",
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <Ionicon
                  beat={false}
                  className="rounded-circle"
                  color="#555"
                  fontSize="50px"
                  icon="logo-yen"
                  rotate={false}
                  shake={false}
                  style={
                                    Object {
                                                      "backgroundColor": "#efefef",
                                                      "padding": "5px",
                                                    }
                  }
/>,
                <p>
                  理财
</p>,
              ],
              "className": "category-item col-3",
              "onClick": [Function],
              "role": "button",
              "style": Object {
                "textAlign": "center",
              },
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "beat": false,
                  "className": "rounded-circle",
                  "color": "#555",
                  "fontSize": "50px",
                  "icon": "logo-yen",
                  "rotate": false,
                  "shake": false,
                  "style": Object {
                    "backgroundColor": "#efefef",
                    "padding": "5px",
                  },
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "理财",
                },
                "ref": null,
                "rendered": "理财",
                "type": "p",
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": "2",
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <Ionicon
                  beat={false}
                  className="rounded-circle"
                  color="#555"
                  fontSize="50px"
                  icon="logo-yen"
                  rotate={false}
                  shake={false}
                  style={
                                    Object {
                                                      "backgroundColor": "#efefef",
                                                      "padding": "5px",
                                                    }
                  }
/>,
                <p>
                  理财
</p>,
              ],
              "className": "category-item col-3",
              "onClick": [Function],
              "role": "button",
              "style": Object {
                "textAlign": "center",
              },
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "beat": false,
                  "className": "rounded-circle",
                  "color": "#555",
                  "fontSize": "50px",
                  "icon": "logo-yen",
                  "rotate": false,
                  "shake": false,
                  "style": Object {
                    "backgroundColor": "#efefef",
                    "padding": "5px",
                  },
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "理财",
                },
                "ref": null,
                "rendered": "理财",
                "type": "p",
              },
            ],
            "type": "div",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
